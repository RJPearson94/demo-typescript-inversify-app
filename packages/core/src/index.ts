import '@src/constant/types';
import '@src/service';
import '@src/inversify.config';
