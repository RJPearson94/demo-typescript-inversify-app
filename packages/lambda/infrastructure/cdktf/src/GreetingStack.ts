import { Construct } from 'constructs';
import crypto from 'crypto';
import fs from 'fs';
import path from 'path';
import { TerraformStack, TerraformOutput } from 'cdktf';
import {
  LambdaFunction,
  CloudwatchLogGroup,
  IamRole,
  LambdaAlias,
  LambdaPermission,
  IamPolicy,
  IamRolePolicyAttachment,
  ApiGatewayRestApi,
  ApiGatewayDeployment,
  ApiGatewayStage,
  ApiGatewayApiKey,
  ApiGatewayUsagePlan,
  ApiGatewayUsagePlanKey
} from '@cdktf/provider-aws';
import openApiTemplate from './templates/openapi.tpl.json';

export type FunctionResources = {
  func: LambdaFunction;
  alias: LambdaAlias;
};

export interface ApiGatewayResources {
  restApi: ApiGatewayRestApi;
  stage: ApiGatewayStage;
  apiKey: ApiGatewayApiKey;
}

export type GreetingStackProps = {
  resourcePrefix?: string;
  api?: {
    stageName?: string;
    quota?: {
      limit?: number;
      period?: string;
    };
  };
  lambda?: {
    alias?: string;
  };
  tags?: { [name: string]: string };
};

export class GreetingStack extends TerraformStack {
  public readonly apiGatewayURL: TerraformOutput;
  public readonly apiKeyArn: TerraformOutput;

  private resourcePrefix: string;

  constructor(scope: Construct, private name: string, private props?: GreetingStackProps) {
    super(scope, name);

    this.resourcePrefix = props?.resourcePrefix ? `${props?.resourcePrefix}-` : '';

    const { alias } = this.createFunction();
    const { stage, apiKey } = this.createAPIGateway(alias);

    new LambdaPermission(this, `api_gateway_invoke_lambda`, {
      statementId: 'AllowAPIGatewayInvoke',
      action: 'lambda:InvokeFunction',
      functionName: alias.arn,
      principal: 'apigateway.amazonaws.com',
      sourceArn: `${stage.executionArn}/*`
    });

    this.apiGatewayURL = new TerraformOutput(this, `api_gateway_url`, {
      value: stage.invokeUrl
    });
    this.apiKeyArn = new TerraformOutput(this, `api_key_arn`, {
      value: apiKey.arn
    });
  }

  private createFunction(): FunctionResources {
    const functionName = `${this.resourcePrefix}${this.name}-function`;

    const logGroup = new CloudwatchLogGroup(this, `lambda_logs`, {
      name: `/aws/lambda/${functionName}`,
      retentionInDays: 7,
      tags: this.props?.tags
    });

    const lambdaRole = new IamRole(this, `lambda_role`, {
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Principal: {
              Service: 'lambda.amazonaws.com'
            },
            Effect: 'Allow'
          }
        ]
      }),
      name: `${functionName}-role`,
      tags: this.props?.tags
    });

    const lambdaPolicy = new IamPolicy(this, `lambda_policy`, {
      name: `${functionName}-policy`,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: ['logs:CreateLogStream', 'logs:PutLogEvents'],
            Resource: logGroup.arn,
            Effect: 'Allow'
          },
          {
            Action: ['xray:PutTraceSegments', 'xray:PutTelemetryRecords'],
            Resource: '*',
            Effect: 'Allow'
          }
        ]
      })
    });

    new IamRolePolicyAttachment(this, `lambda_policy_attachment`, {
      role: lambdaRole.name,
      policyArn: lambdaPolicy.arn
    });

    const lambdaArtefactPath = '../../../dist/lambda.zip';
    const fileContents = fs.readFileSync(path.resolve(__dirname, lambdaArtefactPath));

    const func = new LambdaFunction(this, `lambda`, {
      filename: lambdaArtefactPath,
      sourceCodeHash: crypto.createHash('sha256').update(fileContents).digest('base64'),

      functionName,
      role: lambdaRole.arn,

      runtime: 'nodejs14.x',
      handler: 'main.handler',
      memorySize: 128,
      timeout: 5,
      tags: this.props?.tags,

      tracingConfig: [
        {
          mode: 'PassThrough'
        }
      ]
    });

    const alias = new LambdaAlias(this, `lambda_alias`, {
      name: this.props?.lambda?.alias || 'stable',
      functionName: func.functionName,
      functionVersion: func.version
    });

    return {
      func,
      alias
    };
  }

  private createAPIGateway(alias: LambdaAlias): ApiGatewayResources {
    const template = JSON.stringify(openApiTemplate).replace('${lambda_excution_uri}', alias.invokeArn);

    const apiGatewayName = `${this.resourcePrefix}${this.name}`;

    const restApi = new ApiGatewayRestApi(this, `rest_api`, {
      name: `${apiGatewayName}-rest-api`,
      body: template,
      tags: this.props?.tags
    });

    const deployment = new ApiGatewayDeployment(this, `deployment`, {
      restApiId: restApi.id,

      triggers: {
        redeployment: crypto.createHash('sha512').update(template).digest('hex')
      },

      lifecycle: {
        createBeforeDestroy: true
      }
    });

    const stage = new ApiGatewayStage(this, `stage`, {
      stageName: this.props?.api?.stageName || 'test',
      restApiId: restApi.id,
      deploymentId: deployment.id,
      xrayTracingEnabled: true,
      tags: this.props?.tags
    });

    const apiKey = new ApiGatewayApiKey(this, `api_key`, {
      name: `${apiGatewayName}-api-key`,
      tags: this.props?.tags
    });

    const usagePlan = new ApiGatewayUsagePlan(this, `usage_plan`, {
      name: `${apiGatewayName}-usage-plan`,
      apiStages: [
        {
          apiId: restApi.id,
          stage: stage.stageName
        }
      ],
      quotaSettings: [
        {
          limit: 100,
          period: 'MONTH',
          ...(this.props?.api?.quota || {})
        }
      ],
      tags: this.props?.tags
    });

    new ApiGatewayUsagePlanKey(this, `usage_plan_key`, {
      keyId: apiKey.id,
      usagePlanId: usagePlan.id,
      keyType: 'API_KEY'
    });

    return {
      restApi,
      stage,
      apiKey
    };
  }
}
