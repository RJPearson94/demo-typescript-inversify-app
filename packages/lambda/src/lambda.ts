import 'reflect-metadata';

import container from '@src/inversify.config';
import TYPES from '@src/constant/types';
import GreetingController from '@src/controller/greeting';

const greetingController: GreetingController = container.get<GreetingController>(TYPES.GreetingController);

exports.handler = function(event: any, context: any, callback: Function): void {
  const helloResponse = greetingController.greet();
  callback(null, {
    statusCode: 200,
    body: JSON.stringify({
      message: helloResponse
    })
  });
};
