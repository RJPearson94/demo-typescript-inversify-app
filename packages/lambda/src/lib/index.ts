export * from './apiResponse';
export * from './messageResponse';
export * from './context';