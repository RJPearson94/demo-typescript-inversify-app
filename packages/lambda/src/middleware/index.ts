export * from './inversify';
export * from './lambdaProxy';
export * from './types';
