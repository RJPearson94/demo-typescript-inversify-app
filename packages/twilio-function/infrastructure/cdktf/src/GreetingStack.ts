import { Construct } from 'constructs';
import { TerraformStack, TerraformOutput } from 'cdktf';
import { ServerlessBuild, ServerlessDeployment, ServerlessEnvironment, ServerlessFunction, ServerlessService } from '../.gen/providers/twilio';
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

export type GreetingStackProps = {
  resourceSuffix?: string;
};

export class GreetingStack extends TerraformStack {
  public readonly functionURL: TerraformOutput;

  private resourceSuffix: string;

  constructor(scope: Construct, private name: string, props?: GreetingStackProps) {
    super(scope, name);

    this.resourceSuffix = props?.resourceSuffix ? `-${props?.resourceSuffix}` : '';

    const service = this.createService();
    const func = this.createFunction(service);
    const build = this.createBuild(service, func);
    const environment = this.createEnvironment(service);
    this.createDeployment(service, environment, build);

    this.functionURL = new TerraformOutput(this, `function_url`, {
      value: `https://${environment.domainName}${func.path}`
    });
  }

  private createService(): ServerlessService {
    return new ServerlessService(this, 'service', {
      uniqueName: `${this.name}${this.resourceSuffix}`,
      friendlyName: `${this.name}${this.resourceSuffix}`,
      includeCredentials: false,
      uiEditable: false
    });
  }

  private createFunction(service: ServerlessService): ServerlessFunction {
    const source = '../../../dist/index.js';
    const fileContents = fs.readFileSync(path.resolve(__dirname, source));

    return new ServerlessFunction(this, 'function', {
      serviceSid: service.sid,
      friendlyName: `${this.name}${this.resourceSuffix}`,

      source,
      sourceHash: crypto.createHash('sha256').update(fileContents).digest('base64'),

      contentType: 'application/javascript',
      path: '/inversify',
      visibility: 'public'
    });
  }

  private createBuild(service: ServerlessService, func: ServerlessFunction): ServerlessBuild {
    return new ServerlessBuild(this, 'build', {
      serviceSid: service.sid,

      functionVersion: [
        {
          sid: func.latestVersionSid
        }
      ],

      runtime: 'node12',

      dependencies: {
        twilio: '3.57.0',
        fs: '0.0.1-security',
        lodash: '4.17.21',
        util: '0.12.3',
        xmldom: '0.4.0',
        '@twilio/runtime-handler': '1.0.1'
      },

      polling: [
        {
          enabled: true
        }
      ],

      lifecycle: {
        createBeforeDestroy: true
      }
    });
  }

  private createEnvironment(service: ServerlessService): ServerlessEnvironment {
    return new ServerlessEnvironment(this, 'environment', {
      serviceSid: service.sid,
      uniqueName: `dev${this.resourceSuffix}`
    });
  }

  private createDeployment(service: ServerlessService, environment: ServerlessEnvironment, build: ServerlessBuild): ServerlessDeployment {
    return new ServerlessDeployment(this, 'deployment', {
      serviceSid: service.sid,
      environmentSid: environment.sid,
      buildSid: build.sid,

      lifecycle: {
        createBeforeDestroy: true
      }
    });
  }
}
