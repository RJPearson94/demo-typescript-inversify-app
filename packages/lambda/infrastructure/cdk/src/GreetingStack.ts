import { CfnOutput, Construct, Duration, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';
import { Function, Runtime, Code, Tracing, Alias, IFunction, IAlias } from '@aws-cdk/aws-lambda';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { IRestApi, SpecRestApi, ApiDefinition, RateLimitedApiKey, Period, IApiKey } from '@aws-cdk/aws-apigateway';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import openApiTemplate from './templates/openapi.tpl.json';

export interface FunctionResources {
  func: IFunction;
  alias: IAlias;
}

export interface ApiGatewayResources {
  restApi: IRestApi;
  apiKey: IApiKey;
}

export type GreetingStackProps = {
  resourcePrefix?: string;
  api?: {
    stageName?: string;
    quota?: {
      limit?: number;
      period?: Period;
    };
  };
  lambda?: {
    alias?: string;
  };
} & StackProps;

export class GreetingStack extends Stack {
  public readonly apiGatewayURL: CfnOutput;
  public readonly apiKeyArn: CfnOutput;

  private resourcePrefix: string;

  public constructor(scope: Construct, private id: string, private props?: GreetingStackProps) {
    super(scope, id, props);

    this.resourcePrefix = props?.resourcePrefix ? `${props?.resourcePrefix}-` : '';

    const { alias } = this.createFunction();
    const { restApi, apiKey } = this.createAPIGateway(alias);

    alias.addPermission('AllowApiGatewayInvoke', {
      principal: new ServicePrincipal('apigateway.amazonaws.com'),
      action: 'lambda:InvokeFunction',
      sourceArn: restApi.deploymentStage.restApi.arnForExecuteApi()
    });

    this.apiGatewayURL = new CfnOutput(this, `APIGatewayURL`, {
      value: restApi.deploymentStage.urlForPath()
    });
    this.apiKeyArn = new CfnOutput(this, `APIKeyARN`, {
      value: apiKey.keyArn
    });
  }

  private createFunction(): FunctionResources {
    const functionName = `${this.resourcePrefix}${this.id}`;

    const logGroup = new LogGroup(this, `FunctionLogGroup`, {
      logGroupName: `/aws/lambda/${functionName}`,
      retention: RetentionDays.ONE_WEEK,
      removalPolicy: RemovalPolicy.DESTROY
    });

    const func = new Function(this, `Function`, {
      code: Code.fromAsset('../../dist/lambda.zip'),
      functionName,
      handler: 'main.handler',
      runtime: Runtime.NODEJS_14_X,
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(5)
    });

    logGroup.grantWrite(func);

    const alias = new Alias(this, `FunctionAlias`, {
      aliasName: this.props?.lambda?.alias || 'stable',
      version: func.currentVersion
    });

    return {
      func,
      alias
    };
  }

  private createAPIGateway(alias: IAlias): ApiGatewayResources {
    const template = JSON.stringify(openApiTemplate).replace('${lambda_arn}', alias.functionArn);

    const apiGatewayName = `${this.resourcePrefix}${this.id}`;

    const restApi = new SpecRestApi(this, `ApiGateway`, {
      apiDefinition: ApiDefinition.fromInline(JSON.parse(template)),
      restApiName: `${apiGatewayName}-rest-api`,
      deploy: true,
      cloudWatchRole: false,
      deployOptions: {
        tracingEnabled: true,
        stageName: this.props?.api?.stageName || 'test'
      }
    });

    const apiKey = new RateLimitedApiKey(this, `ApiKey`, {
      apiKeyName: `${apiGatewayName}-api-key`,
      resources: [restApi],
      apiStages: [
        {
          api: restApi,
          stage: restApi.deploymentStage
        }
      ],
      quota: {
        limit: 100,
        period: Period.MONTH,
        ...(this.props?.api?.quota || {})
      }
    });

    return {
      restApi,
      apiKey
    };
  }
}
