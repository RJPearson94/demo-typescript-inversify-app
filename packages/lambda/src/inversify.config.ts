import { container } from '@rjpearson94/core';

import { TYPES } from '@src/constant/types';
import { GreetingController, GreetingControllerImpl } from '@src/controller';

container.bind<GreetingController>(TYPES.GreetingController).to(GreetingControllerImpl);

export { container };
