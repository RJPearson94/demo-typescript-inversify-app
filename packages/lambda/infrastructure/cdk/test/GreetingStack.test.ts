import { SynthUtils } from '@aws-cdk/assert';
import { App } from '@aws-cdk/core';

import { GreetingStack } from '../src/GreetingStack';

test('create greeting stack', () => {
  const app = new App();
  const stack = new GreetingStack(app, 'greeting-stack');
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
