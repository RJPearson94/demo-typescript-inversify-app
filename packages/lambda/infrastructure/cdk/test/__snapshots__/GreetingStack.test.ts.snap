// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create greeting stack 1`] = `
Object {
  "Outputs": Object {
    "APIGatewayURL": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiGateway11E7F47B",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiGatewayDeploymentStagetest6EE49CA4",
            },
            "/",
          ],
        ],
      },
    },
    "APIKeyARN": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "arn:",
            Object {
              "Ref": "AWS::Partition",
            },
            ":apigateway:",
            Object {
              "Ref": "AWS::Region",
            },
            "::/apikeys/",
            Object {
              "Ref": "ApiKey25876BBE",
            },
          ],
        ],
      },
    },
    "ApiGatewayEndpoint5AA8EC3A": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiGateway11E7F47B",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiGatewayDeploymentStagetest6EE49CA4",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6ArtifactHash2E11E2F3": Object {
      "Description": "Artifact hash for asset \\"2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6\\"",
      "Type": "String",
    },
    "AssetParameters2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6S3Bucket49BAD129": Object {
      "Description": "S3 bucket for asset \\"2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6\\"",
      "Type": "String",
    },
    "AssetParameters2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6S3VersionKey2E9334C6": Object {
      "Description": "S3 key for asset version \\"2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ApiGateway11E7F47B": Object {
      "Properties": Object {
        "Body": Object {
          "components": Object {
            "schemas": Object {
              "GreetingResponse": Object {
                "properties": Object {
                  "message": Object {
                    "type": "string",
                  },
                },
                "type": "object",
              },
            },
            "securitySchemes": Object {
              "ApiKey": Object {
                "in": "header",
                "name": "x-api-key",
                "type": "apiKey",
              },
            },
          },
          "info": Object {
            "title": "Greeting API",
            "version": "1.0.0",
          },
          "openapi": "3.0.1",
          "paths": Object {
            "/v1/greet": Object {
              "get": Object {
                "description": "Provide Greeting Message when lambda is invoked. Lambda is written in TypeScript",
                "operationId": "getV1Greet",
                "responses": Object {
                  "200": Object {
                    "content": Object {
                      "application/json": Object {
                        "schema": Object {
                          "$ref": "#/components/schemas/GreetingResponse",
                        },
                      },
                    },
                    "description": "Successfully Greet",
                  },
                },
                "security": Array [
                  Object {
                    "ApiKey": Array [],
                  },
                ],
                "summary": "Provide Greeting Message",
                "x-amazon-apigateway-integration": Object {
                  "httpMethod": "POST",
                  "responses": Object {
                    "default": Object {
                      "statusCode": "200",
                    },
                  },
                  "timeoutInMillis": 10000,
                  "type": "aws_proxy",
                  "uri": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:aws:apigateway:\${AWS::Region}:lambda:path/2015-03-31/functions/",
                        Object {
                          "Ref": "FunctionAliasE9E29399",
                        },
                        "/invocations",
                      ],
                    ],
                  },
                },
              },
            },
          },
        },
        "Name": "greeting-stack-rest-api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ApiGatewayDeploymentA26796E812aabc4e95265843f7f6d8fb8aec3968": Object {
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "ApiGateway11E7F47B",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiGatewayDeploymentStagetest6EE49CA4": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "ApiGatewayDeploymentA26796E812aabc4e95265843f7f6d8fb8aec3968",
        },
        "RestApiId": Object {
          "Ref": "ApiGateway11E7F47B",
        },
        "StageName": "test",
        "TracingEnabled": true,
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiKey25876BBE": Object {
      "Properties": Object {
        "Enabled": true,
        "Name": "greeting-stack-api-key",
        "StageKeys": Array [
          Object {
            "RestApiId": Object {
              "Ref": "ApiGateway11E7F47B",
            },
            "StageName": Object {
              "Ref": "ApiGatewayDeploymentStagetest6EE49CA4",
            },
          },
        ],
      },
      "Type": "AWS::ApiGateway::ApiKey",
    },
    "ApiKeyUsagePlanResourceDB236EC3": Object {
      "Properties": Object {
        "ApiStages": Array [
          Object {
            "ApiId": Object {
              "Ref": "ApiGateway11E7F47B",
            },
            "Stage": Object {
              "Ref": "ApiGatewayDeploymentStagetest6EE49CA4",
            },
            "Throttle": Object {},
          },
        ],
        "Quota": Object {
          "Limit": 100,
          "Period": "MONTH",
        },
      },
      "Type": "AWS::ApiGateway::UsagePlan",
    },
    "ApiKeyUsagePlanResourceUsagePlanKeyResourceD56B1068": Object {
      "Properties": Object {
        "KeyId": Object {
          "Ref": "ApiKey25876BBE",
        },
        "KeyType": "API_KEY",
        "UsagePlanId": Object {
          "Ref": "ApiKeyUsagePlanResourceDB236EC3",
        },
      },
      "Type": "AWS::ApiGateway::UsagePlanKey",
    },
    "Function76856677": Object {
      "DependsOn": Array [
        "FunctionServiceRoleDefaultPolicy2F49994A",
        "FunctionServiceRole675BB04A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6S3Bucket49BAD129",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6S3VersionKey2E9334C6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters2cb65ca7825cd6e957d8c0ddac8735a605fa5867c0f65c8152c10bd2c5b1d8f6S3VersionKey2E9334C6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "greeting-stack",
        "Handler": "main.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "FunctionServiceRole675BB04A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 5,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionAliasAllowApiGatewayInvokeD66B39EA": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Ref": "FunctionAliasE9E29399",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApiGateway11E7F47B",
              },
              "/*/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "FunctionAliasE9E29399": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "Function76856677",
        },
        "FunctionVersion": Object {
          "Fn::GetAtt": Array [
            "FunctionCurrentVersion4E2B2261ebb20bcf35ea65f29e463ddf4c4f5b3e",
            "Version",
          ],
        },
        "Name": "stable",
      },
      "Type": "AWS::Lambda::Alias",
    },
    "FunctionCurrentVersion4E2B2261ebb20bcf35ea65f29e463ddf4c4f5b3e": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "Function76856677",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "FunctionLogGroupBD1576D5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": "/aws/lambda/greeting-stack",
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "FunctionServiceRole675BB04A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FunctionServiceRoleDefaultPolicy2F49994A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "FunctionLogGroupBD1576D5",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FunctionServiceRoleDefaultPolicy2F49994A",
        "Roles": Array [
          Object {
            "Ref": "FunctionServiceRole675BB04A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
